storage/remote/azuread/token_provider.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2023 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package azuread
+
+import (
+	"context"
+	"errors"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+)
+
+// tokenProvider is used to store and retrieve Azure AD accessToken
+type tokenProvider struct {
+	// token is member used to store the current valid accessToken
+	token string
+	ctx   context.Context
+	// refreshDuration is used to store the refresh time duration of the current valid accessToken
+	refreshDuration time.Duration
+	// credentialClient is the Azure AD credential client that is being used to retrive accessToken
+	credentialClient azcore.TokenCredential
+	options          *policy.TokenRequestOptions
+}
+
+// TokenProvider is the interface for Credential client
+type TokenProvider interface {
+	GetAccessToken() string
+}
+
+// NewTokenProvider helps to fetch accessToken for different types of credential. This also takes care of
+// refreshing the accessToken before expiry. This accessToken is attached to the Authorization header while making requests.
+func NewTokenProvider(ctx context.Context, cfg *AzureAdConfig, cred azcore.TokenCredential) (TokenProvider, error) {
```

Comment:
Same thing with the private fields/methods here.

Reply:
made them private